diff --git a/drivers/iommu/vsi-iommu.c b/drivers/iommu/vsi-iommu.c
index 42c424496..304c98a9c 100644
--- a/drivers/iommu/vsi-iommu.c
+++ b/drivers/iommu/vsi-iommu.c
@@ -6,13 +6,6 @@
  * Module Authors:	Yandong Lin <yandong.lin@rock-chips.com>
  *			Simon Xue <xxm@rock-chips.com>
  *			Benjamin Gaignard <benjamin.gaignard@collabora.com>
- *
- * This hardware block is using a 2 pages tables allocation structure.
- * That make very similar to Rockhip iommu hardware blocks but it has
- * it own driver because the registers offset and configuration bits
- * are completely different. An additional reason is that this hardware
- * has been developed by Verisilicon to be used by their hardware video
- * decoders and not for a general purpose like Rockchip iommus.
  */
 
 #include <linux/clk.h>
@@ -35,9 +28,11 @@
 #include <linux/pm_runtime.h>
 #include <linux/slab.h>
 #include <linux/spinlock.h>
+#include <linux/vsi-iommu.h>
 
 #include "iommu-pages.h"
 
+
 struct vsi_iommu {
 	struct device *dev;
 	void __iomem *regs;
@@ -48,7 +43,6 @@ struct vsi_iommu {
 	struct iommu_domain *domain; /* domain to which iommu is attached */
 	spinlock_t lock; /* lock to protect vsi_iommu fields */
 	int irq;
-	bool enable;
 };
 
 struct vsi_iommu_domain {
@@ -66,6 +60,7 @@ static struct iommu_domain vsi_identity_domain;
 
 #define NUM_DT_ENTRIES	1024
 #define NUM_PT_ENTRIES	1024
+#define PT_SIZE		(NUM_PT_ENTRIES * sizeof(u32))
 
 #define SPAGE_SIZE	BIT(12)
 
@@ -189,6 +184,35 @@ static u32 vsi_iova_page_offset(u32 iova)
 	return (iova & VSI_IOVA_PAGE_MASK) >> VSI_IOVA_PAGE_SHIFT;
 }
 
+static void vsi_iommu_flush_tlb_all(struct iommu_domain *domain)
+{
+	struct vsi_iommu_domain *vsi_domain = to_vsi_domain(domain);
+	struct list_head *pos;
+	unsigned long flags;
+
+	spin_lock_irqsave(&vsi_domain->lock, flags);
+
+	list_for_each(pos, &vsi_domain->iommus) {
+		struct vsi_iommu *iommu;
+		int ret;
+
+		iommu = list_entry(pos, struct vsi_iommu, node);
+		ret = pm_runtime_resume_and_get(iommu->dev);
+		if (ret < 0)
+			continue;
+
+		spin_lock(&iommu->lock);
+
+		writel(VSI_MMU_BIT_FLUSH, iommu->regs + VSI_MMU_FLUSH_BASE);
+		writel(0, iommu->regs + VSI_MMU_FLUSH_BASE);
+
+		spin_unlock(&iommu->lock);
+		pm_runtime_put_autosuspend(iommu->dev);
+	}
+
+	spin_unlock_irqrestore(&vsi_domain->lock, flags);
+}
+
 static irqreturn_t vsi_iommu_irq(int irq, void *dev_id)
 {
 	struct vsi_iommu *iommu = dev_id;
@@ -363,28 +387,6 @@ static int vsi_iommu_map_iova(struct vsi_iommu_domain *vsi_domain, u32 *pte_addr
 	return 0;
 }
 
-static void vsi_iommu_flush_tlb(struct iommu_domain *domain)
-{
-	struct vsi_iommu_domain *vsi_domain = to_vsi_domain(domain);
-	struct vsi_iommu *iommu;
-
-	list_for_each_entry(iommu, &vsi_domain->iommus, node) {
-		if (pm_runtime_get(iommu->dev) < 0)
-			continue;
-
-		spin_lock(&iommu->lock);
-
-		if (iommu->enable) {
-			writel(VSI_MMU_BIT_FLUSH, iommu->regs + VSI_MMU_FLUSH_BASE);
-			writel(0, iommu->regs + VSI_MMU_FLUSH_BASE);
-		}
-
-		spin_unlock(&iommu->lock);
-
-		pm_runtime_put_autosuspend(iommu->dev);
-	}
-}
-
 static size_t vsi_iommu_unmap(struct iommu_domain *domain, unsigned long _iova,
 			      size_t size, size_t count, struct iommu_iotlb_gather *gather)
 {
@@ -407,10 +409,7 @@ static size_t vsi_iommu_unmap(struct iommu_domain *domain, unsigned long _iova,
 	pte_addr = (u32 *)phys_to_virt(pt_phys) + vsi_iova_pte_index(iova);
 	pte_dma = pt_phys + vsi_iova_pte_index(iova) * sizeof(u32);
 	unmap_size = vsi_iommu_unmap_iova(vsi_domain, pte_addr, pte_dma, size);
-	if (!unmap_size)
-		goto unlock;
 
-	vsi_iommu_flush_tlb(domain);
 unlock:
 	spin_unlock_irqrestore(&vsi_domain->lock, flags);
 
@@ -482,9 +481,6 @@ static int vsi_iommu_map(struct iommu_domain *domain, unsigned long _iova,
 				 paddr, size, prot);
 	if (!ret)
 		*mapped = size;
-
-	vsi_iommu_flush_tlb(domain);
-
 	spin_unlock_irqrestore(&vsi_domain->lock, flags);
 
 	return ret;
@@ -493,14 +489,12 @@ static int vsi_iommu_map(struct iommu_domain *domain, unsigned long _iova,
 static void vsi_iommu_disable(struct vsi_iommu *iommu)
 {
 	writel(0, iommu->regs + VSI_MMU_AHB_CONTROL_BASE);
-	iommu->enable = false;
 }
 
 static int vsi_iommu_identity_attach(struct iommu_domain *domain,
 				     struct device *dev, struct iommu_domain *old)
 {
 	struct vsi_iommu *iommu = dev_iommu_priv_get(dev);
-	struct vsi_iommu_domain *vsi_domain = to_vsi_domain(domain);
 	unsigned long flags;
 	int ret;
 
@@ -508,8 +502,7 @@ static int vsi_iommu_identity_attach(struct iommu_domain *domain,
 	if (ret < 0)
 		return ret;
 
-	spin_lock_irqsave(&vsi_domain->lock, flags);
-	spin_lock(&iommu->lock);
+	spin_lock_irqsave(&iommu->lock, flags);
 	if (iommu->domain == domain)
 		goto unlock;
 
@@ -519,8 +512,7 @@ static int vsi_iommu_identity_attach(struct iommu_domain *domain,
 	iommu->domain = domain;
 
 unlock:
-	spin_unlock(&iommu->lock);
-	spin_unlock_irqrestore(&vsi_domain->lock, flags);
+	spin_unlock_irqrestore(&iommu->lock, flags);
 	pm_runtime_put_autosuspend(iommu->dev);
 	return 0;
 }
@@ -545,15 +537,41 @@ static void vsi_iommu_enable(struct vsi_iommu *iommu, struct iommu_domain *domai
 	writel(VSI_MMU_OUT_OF_BOUND, iommu->regs + VSI_MMU_CONFIG1_BASE);
 	writel(VSI_MMU_BIT_ENABLE, iommu->regs + VSI_MMU_AHB_EXCEPTION_BASE);
 	writel(VSI_MMU_BIT_ENABLE, iommu->regs + VSI_MMU_AHB_CONTROL_BASE);
-	iommu->enable = true;
 }
 
+void vsi_iommu_restore_ctx(struct iommu_domain *domain)
+{
+	struct vsi_iommu_domain *vsi_domain = to_vsi_domain(domain);
+	struct list_head *pos;
+	unsigned long flags;
+
+	spin_lock_irqsave(&vsi_domain->lock, flags);
+
+	list_for_each(pos, &vsi_domain->iommus) {
+		struct vsi_iommu *iommu;
+
+		iommu = list_entry(pos, struct vsi_iommu, node);
+		if (!iommu)
+			continue;
+
+		spin_lock(&iommu->lock);
+
+		writel(VSI_MMU_BIT_FLUSH, iommu->regs + VSI_MMU_FLUSH_BASE);
+		writel(0, iommu->regs + VSI_MMU_FLUSH_BASE);
+
+		spin_unlock(&iommu->lock);
+	}
+
+	spin_unlock_irqrestore(&vsi_domain->lock, flags);
+}
+EXPORT_SYMBOL_GPL(vsi_iommu_restore_ctx);
+
 static int vsi_iommu_attach_device(struct iommu_domain *domain,
 				   struct device *dev, struct iommu_domain *old)
 {
 	struct vsi_iommu *iommu = dev_iommu_priv_get(dev);
 	struct vsi_iommu_domain *vsi_domain = to_vsi_domain(domain);
-	unsigned long flags;
+	unsigned long flags, flags2;
 	int ret = 0;
 
 	ret = pm_runtime_resume_and_get(iommu->dev);
@@ -561,7 +579,7 @@ static int vsi_iommu_attach_device(struct iommu_domain *domain,
 		return ret;
 
 	spin_lock_irqsave(&vsi_domain->lock, flags);
-	spin_lock(&iommu->lock);
+	spin_lock_irqsave(&iommu->lock, flags2);
 
 	vsi_iommu_enable(iommu, domain);
 	writel(VSI_MMU_BIT_FLUSH, iommu->regs + VSI_MMU_FLUSH_BASE);
@@ -572,7 +590,7 @@ static int vsi_iommu_attach_device(struct iommu_domain *domain,
 
 	iommu->domain = domain;
 
-	spin_unlock(&iommu->lock);
+	spin_unlock_irqrestore(&iommu->lock, flags2);
 	spin_unlock_irqrestore(&vsi_domain->lock, flags);
 	pm_runtime_put_autosuspend(iommu->dev);
 	return ret;
@@ -653,6 +671,7 @@ static const struct iommu_ops vsi_iommu_ops = {
 		.attach_dev		= vsi_iommu_attach_device,
 		.map_pages		= vsi_iommu_map,
 		.unmap_pages		= vsi_iommu_unmap,
+		.flush_iotlb_all	= vsi_iommu_flush_tlb_all,
 		.iova_to_phys		= vsi_iommu_iova_to_phys,
 		.free			= vsi_iommu_domain_free,
 	}
@@ -708,8 +727,7 @@ static int vsi_iommu_probe(struct platform_device *pdev)
 	pm_runtime_use_autosuspend(dev);
 	pm_runtime_enable(dev);
 
-	err = iommu_device_sysfs_add(&iommu->iommu, dev, NULL, "%s",
-				     dev_name(dev));
+	err = iommu_device_sysfs_add(&iommu->iommu, dev, NULL, dev_name(dev));
 	if (err)
 		goto err_runtime_disable;
 
@@ -750,7 +768,7 @@ static int __maybe_unused vsi_iommu_suspend(struct device *dev)
 static int __maybe_unused vsi_iommu_resume(struct device *dev)
 {
 	struct vsi_iommu *iommu = dev_get_drvdata(dev);
-	unsigned long flags;
+	unsigned long flags, flags2;
 	int ret;
 
 	ret = clk_bulk_enable(iommu->num_clocks, iommu->clocks);
@@ -761,9 +779,9 @@ static int __maybe_unused vsi_iommu_resume(struct device *dev)
 		struct vsi_iommu_domain *vsi_domain = to_vsi_domain(iommu->domain);
 
 		spin_lock_irqsave(&vsi_domain->lock, flags);
-		spin_lock(&iommu->lock);
+		spin_lock_irqsave(&iommu->lock, flags2);
 		vsi_iommu_enable(iommu, iommu->domain);
-		spin_unlock(&iommu->lock);
+		spin_unlock_irqrestore(&iommu->lock, flags2);
 		spin_unlock_irqrestore(&vsi_domain->lock, flags);
 	}
 
diff --speed-large-files --no-dereference --minimal -Naur linux-7.0-rc1/include/linux/vsi-iommu.h linux-7.0-rc1/include/linux/vsi-iommu.h
--- linux-7.0-rc1/include/linux/vsi-iommu.h	1970-01-01 01:00:00.000000000 +0100
++ linux-7.0-rc1/include/linux/vsi-iommu.h	2026-03-02 12:41:56.770149059 +0100
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * verisilicon iommu: simple virtual address space management
+ *
+ * Copyright (c) 2025, Collabora
+ *
+ * Written by Benjamin Gaignard <benjamin.gaignard@collabora.com>
+ */
+
+#ifndef _VSI_IOMMU_H_
+#define _VSI_IOMMU_H_
+
+struct iommu_domain;
+
+#if IS_ENABLED(CONFIG_VSI_IOMMU)
+void vsi_iommu_restore_ctx(struct iommu_domain *domain);
+#else
+static inline void vsi_iommu_restore_ctx(struct iommu_domain *domain) {}
+#endif
+
+#endif
